File: analytics/R/Readhdf.R
# 
# Processing the weather HDF5 files in R
#
# This script opens an HDF5 file and makes a plot of day/night 
# average temperature from the "daily" dataset contained inside
# the file.
#

# Required libraries to read and process the HDF5 files
library(hdf5r)
library(nanotime)
library(dplyr)
library(ggplot2)

#+
# Change the dayTime indicator from an integer (1=day, 0=night) to a 
# string. It will be converted to a factor when loaded into a data.frame
#-
dayTimeName <- function(x) {
  if (x['daytime']==1)
    return("Day")
  else
    return("Night")
}

#+
# Open the HDF5 file
#-
f0 <- H5File$new(filename = '../data/monthly-201805.hdf5', mode = 'r')

#+
# The dataset block0 contains the dates in interger64 format
#-
d0 <- f0$open('daily/block0_values')
c0 <- f0$open('daily/block0_items')

#+
# The dataset block1 contains the floating point varianles:
# temperature, pressure, light and humidity
#-
d1 <- f0$open('daily/block1_values')
c1 <- f0$open('daily/block1_items')

#+
# The dataset block2 contains the integer variables:
# daytime indicator and count of records
#-
d2 <- f0$open('daily/block2_values')
c2 <- f0$open('daily/block2_items')

#+
# Convert the integer64 nanotimes to POSIXct, based on UTC,
# and put the result into an array
#-
a0 <- as.POSIXct(nanotime(d0$read()),tz='UTC')

#+
# Put the float and integer values into arrays
#-
a1 <- (d1$read())
a2 <- (d2$read())

# +
# Convert the arrays to data.frames and add the column names
# which are in the xxxx_items datasets in the HDF5 file
#-
df0 <- data.frame(a0)
colnames(df0) <- c0$read() 
df1 <- data.frame(t(a1))
colnames(df1) <- c1$read() 
df2 <- data.frame(t(a2))
colnames(df2) <- c2$read() 

#+
# Bind the three data frames to build an unified one with all the
# data.
#-
f <- cbind(df0,df1,df2)

#+
# Convert the datatime indicator to a factor.
# Convert to wide format, selection just the temperature
#-
f <- dplyr::mutate(f,daytime=as.factor(apply(f,1,dayTimeName)))
fw <- reshape(data = f, direction = 'wide', idvar=c('date'),timevar='daytime',
               drop=c('light','pressure','humidity','count'))

#+
# Add a column with the difference between day and night temperatures
# Convert back to long format
#-
fw <- mutate(fw, temperature.Difference = temperature.Day - temperature.Night)
fw2 <- reshape(fw,direction = 'long', varying=c('temperature.Day','temperature.Night','temperature.Difference'))

#+
# Prepare and draw a plot
#-
plt <- ggplot(data = fw2) + geom_line(aes(x=date,y=temperature,group=time,color=time))
plot(plt)

File: analytics/Julia/ReadHDF.jl
#
# Processing the weather HDF5 files in Julia 1.0
#
# This script opens an HDF5 file and creates DataFrame objects for the three
# datasets it contains. It also makes a plot of day/night
# average temperature from the "daily" dataset.
#

using HDF5
using Base.Filesystem
using DataFrames
using CategoricalArrays
using Gadfly
using Dates

#+
# Change this variables to point to the HDF file you want to process
#-
filePath = relpath("../data")
fileName = "monthly-201808.hdf5"

#+
# Open the hdf5 file
#-
h5f = h5open(joinpath(filePath,fileName),"r")

#+
# Extract references to the data and labes for the rawdata dataset
#-
dr0  = h5f["rawdata/block0_values"] # Floats dataset
dr1  = h5f["rawdata/block1_values"] # integer64 (datetime) dataset
dr0n = h5f["rawdata/block0_items"]
dr1n = h5f["rawdata/block1_items"]

#+
# Create DataFrames for the block0 (floats) and block1 (integer64 containing
# times in nanosecods). Note the tick used to transpose the read matrices.
# Build an array with the field labels from both datasets.
#-
dfr0 = DataFrame(read(dr0)')
dfr1 = DataFrame(read(dr1)')
drn  = vcat(read(dr1n),read(dr0n))

#+
# Concatenate both DataFrames and assign the column names.
#-
dfr = hcat(dfr1,dfr0)
names!(dfr,Symbol.(drn))

#+
# Convert the interger64 values to proper times. Remember those
# times are UTC, but the Julia datatime is plain (has no TZ info)
#-
dfr[:time] = unix2datetime.(div.(dfr[:time],1000000000))


#+
# Extract references to the data and labes for the hourly dataset
#-
dh0 =  h5f["hourly/block0_values"] # Floats dataset
dh1  = h5f["hourly/block1_values"] # integer dataset (count)
dh2  = h5f["hourly/block2_values"] # integer64 dataset (time)
dh0n = h5f["hourly/block0_items"]
dh1n = h5f["hourly/block1_items"]
dh2n = h5f["hourly/block2_items"]

#+
# Create dataframes from the dataset contents and prepare an
# array with the field labels
#
dfh0 = DataFrame(read(dh0)')
dfh1 = DataFrame(read(dh1)')
dfh2 = DataFrame(read(dh2)')
dfn = vcat(read(dh0n), read(dh1n), read(dh2n))

#+
# Concatenate the datasets, assign column names and convert int64 to datetime
#-
dfh = hcat(dfh0,dfh1,dfh2)
names!(dfh,Symbol.(dfn))
dfh[:time] = unix2datetime.(div.(dfh[:time],1000000000))

#+
# Get references for the daily datasets (contents and labels)
#-
dd0 = h5f["daily/block0_values"] # Integer64 (datetime)
dd1 = h5f["daily/block1_values"] # Floats
dd2 = h5f["daily/block2_values"] # Integers (count and daylight indicator)
dd0n = h5f["daily/block0_items"]
dd1n = h5f["daily/block1_items"]
dd2n = h5f["daily/block2_items"]

#+
# Create DataFrames and label array
#-
dfd0 = DataFrame(read(dd0)')
dfd1 = DataFrame(read(dd1)')
dfd2 = DataFrame(read(dd2)')
ddn = vcat(read(dd0n), read(dd1n), read(dd2n))

#+
# Concatenate dataframes, add column labels and convert datetime
#-
dfd = hcat(dfd0,dfd1,dfd2)
names!(dfd,Symbol.(ddn))
dfd[:date] = unix2datetime.(div.(dfd[:date],1000000000))

#+
# Convert the integer daytime indicator to a categorical, descriptive
# variable.
#-
dfd[:daytime] = CategoricalArray(map(x-> if x==1 return "Day" else return "Night" end , dfd[:daytime]))

#+
# Unstack the data, extracting just the temperature, so we have the day
# and night temperatures in the same row. Discard days with incomplete
# data and add a day-night difference column.
#-
dfdu = dropmissing!(unstack(dfd,[:date],:daytime,:temperature))
dfdu[:Difference] = dfdu[:Day] - dfdu[:Night]

#+
# Stack again the dataframe so we can use Gadfly to plot it.
#-
dfdt = stack(dfdu,[:Day,:Night,:Difference], variable_name=:daytime, value_name=:temperature)
plot(dfdt,x=:date,y=:temperature,color=:daytime,Geom.line)
